### PR TITLE
Fixed #27 and #28 Memory server profiling and resolved KVS and heap map synchronization issue

### LINKED ISSUE #27: Fix synchronization issues in KVS map and heap map in memoryserver
Memory server uses multiple maps like KVS map and heap maps to fasten the access to heaps and KVS. When multiple threads try to create it, it causes exception in some of the threads.

### LINKED ISSUE #28: Enhance OpenFAM profiling to include datapath and memory server profiles
Add profiling code to enhance datapath (libfabric) and memory server profiling. 
Memory server profiling needs to include metadata and NVMM calls.

### PR DESCRIPTION
**Pull request by gautham-bhat-k**
Fixed #28 Enhance OpenFAM profiling to include datapath and memory server profiles
Fixed #27 Fix synchronization issues in KVS map and heap map in memoryserver

### CODE REVIEW

File: test/microbench/fam-api-mb/fam_microbenchmark_allocator.cpp [hunk 1]
```diff
@@ -0,0 +1,542 @@
+/*
+ * fam_microbenchmark.cpp
```
Comment (mashoodk) [inline test/microbench/fam-api-mb/fam_microbenchmark_allocator.cpp:2]: typo and copyrights can be 2020

File: test/microbench/fam-api-mb/scripts/test_series_allocator.sh [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+#INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#CONTRACT, STRICT LIABILITY, OR TORT(INCLUDING NEGLIGENCE OR OTHERWISE)
+#ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#POSSIBILITY OF SUCH DAMAGE.
+#
+#See https: // spdx.org/licenses/BSD-3-Clause
+#
+#
+
+#!/ bin / bash
```
Comment (mashoodk) [inline test/microbench/fam-api-mb/scripts/test_series_allocator.sh:31]: This has to be in first line

File: test/microbench/fam-api-mb/scripts/test_series_atomic.sh [hunk 3]
```diff
@@ -1,34 +1,34 @@
...
+#CONTRACT, STRICT LIABILITY, OR TORT(INCLUDING NEGLIGENCE OR OTHERWISE)
+#ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#POSSIBILITY OF SUCH DAMAGE.
+#
+#See https: // spdx.org/licenses/BSD-3-Clause
+#
+#
 
-#!/bin/bash
+#!/ bin / bash
```
Comment (mashoodk) [inline test/microbench/fam-api-mb/scripts/test_series_atomic.sh:31]: Move to first line and no spaces

File: test/microbench/fam-api-mb/scripts/run_datapath_mb_CLUSTER.sh [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+ # SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ #    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ # CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ # ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ # POSSIBILITY OF SUCH DAMAGE.
+ #
+ # See https://spdx.org/licenses/BSD-3-Clause
+ #
+ #
+#!/bin/bash -x
```
Comment (mashoodk) [inline test/microbench/fam-api-mb/scripts/run_datapath_mb_CLUSTER.sh:30]: first line

File: src/allocator/memserver_allocator.cpp [hunk 5]
```diff
@@ -721,31 +908,57 @@ int Memserver_Allocator::open_heap(uint64_t regionId) {
...
+            }
             delete heap;
-            throw Memserver_Exception(RBT_HEAP_NOT_INSERTED,
-                                      message.str().c_str());
         }
-
+        NVMM_PROFILE_END_OPS(HeapMapFindOp)
+        // Profile_Time end1 = NVMM_PROFILE_GET_TIME();
+        // Profile_Time total1 = NVMM_PROFILE_TIME_DIFF_NS(start2, end1);
+        // NVMM_PROFILE_ADD_TO_TOTAL_OPS(HeapMapFindOp, total1);
```
Comment (VrashiPonnappa) [inline src/allocator/memserver_allocator.cpp:956]: We should move mutex unlock before close, to reduce the hold time of lock.

**Pull request merged**